relay-server/src/services/processor.rs

```diff
@@ -1903,47 +1903,74 @@ impl EnvelopeProcessorService {
     /// Check and apply rate limits to metrics buckets.
     #[cfg(feature = "processing")]
     fn handle_rate_limit_buckets(&self, message: RateLimitBuckets) {
+        relay_log::trace!("handle_rate_limit_buckets");
         let RateLimitBuckets { mut bucket_limiter } = message;
 
         let scoping = *bucket_limiter.scoping();
 
         if let Some(rate_limiter) = self.inner.rate_limiter.as_ref() {
-            let item_scoping = relay_quotas::ItemScoping {
-                category: DataCategory::Transaction,
-                scoping: &scoping,
-                namespace: None,
-            };
-
             let global_config = self.inner.global_config.current();
             let quotas = DynamicQuotas::new(&global_config, bucket_limiter.quotas());
 
             // We set over_accept_once such that the limit is actually reached, which allows subsequent
             // calls with quantity=0 to be rate limited.
             let over_accept_once = true;
-            let rate_limits = rate_limiter.is_rate_limited(
-                quotas,
-                item_scoping,
-                bucket_limiter.transaction_count(),
-                over_accept_once,
-            );
 
-            let was_enforced = bucket_limiter.enforce_limits(
-                rate_limits.as_ref().map_err(|_| ()),
-                self.inner.outcome_aggregator.clone(),
-            );
+            let transaction_rate_limits = bucket_limiter.transaction_count().and_then(|count| {
+                rate_limiter
+                    .is_rate_limited(
+                        quotas,
+                        ItemScoping {
+                            category: DataCategory::Transaction,
+                            scoping: &scoping,
+                            namespace: None,
+                        },
+                        count,
+                        over_accept_once,
+                    )
+                    .ok() // don't limit if there was a redis error
+            });
 
-            if was_enforced {
-                if let Ok(limits) = rate_limits {
+            let span_rate_limits = bucket_limiter.span_count().and_then(|count| {
+                rate_limiter
+                    .is_rate_limited(
+                        quotas,
+                        ItemScoping {
+                            category: DataCategory::Span,
+                            scoping: &scoping,
+                            namespace: None,
+                        },
+                        count,
+                        over_accept_once,
+                    )
+                    .ok() // don't limit if there was a redis error
+            });
+
+            let merged_rate_limits = match (transaction_rate_limits, span_rate_limits) {
```

Comment:
Generic merge of rate limits, which doesn't need to know about transactions/spans, make it a method on the result?